[stable/nginx-ingress] Fixes compatibility with 1.16

null93 opened the pull request:
Signed-off-by: Rafael Grigorian <me@raffi.io>

<!--
Thank you for contributing to helm/charts. Before you submit this PR we'd like to
make sure you are aware of our technical requirements and best practices:

* https://github.com/helm/charts/blob/master/CONTRIBUTING.md#technical-requirements
* https://github.com/helm/helm/tree/master/docs/chart_best_practices

For a quick overview across what we will look at reviewing your PR, please read
our review guidelines:

* https://github.com/helm/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and
help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the
history. This will make it easier to identify new changes. The PR will be squashed
anyways when it is merged. Thanks.

For fast feedback, please @-mention maintainers that are listed in the Chart.yaml file.

Please make sure you test your changes before you push them. Once pushed, a CircleCI
will run across your changes and do some initial checks and linting. These checks run
very quickly. Please check the results. We would like these checks to pass before we
even continue reviewing your changes.
-->

#### What this PR does / why we need it:

Makes stable/nginx-ingress compatible with 1.16. A quick overview of deprecated APIs can be found on their blog post: https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16/

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/chart]`)

taharah commented on stable/nginx-ingress/templates/controller-daemonset.yaml:
```diff
@@ -1,7 +1,7 @@
 {{- $useHostPort := .Values.controller.daemonset.useHostPort -}}
 {{- $hostPorts := .Values.controller.daemonset.hostPorts -}}
-apiVersion: extensions/v1beta1
+apiVersion: apps/v1
```
I think this could potentially be a breaking change for some users, so can we just make this configurable with the default being the existing API version?

null93 replied:
The **DaemonSet** resource is no longer available in `v1.16` under the `extensions/v1beta1 ` API. It is now only available in the `apps/v1` API. This was available in the `apps/v1` API since `v1.9`. This should be backward compatible as long as `.spec.selector` is explicitly defined within the **DaemonSet** resource.

Source: https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16/

taharah commented on stable/nginx-ingress/templates/controller-daemonset.yaml:
```diff
@@ -12,6 +12,14 @@ metadata:
+      component: "{{ .Values.controller.name }}"
+      release: {{ .Release.Name }}
+      {{- if .Values.controller.podLabels }}
+{{ toYaml .Values.controller.podLabels | indent 6}}
```
Let's not add the podLabels to the selector since this field is immutable.

null93 replied:
According to [https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/#pod-selector](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/#pod-selector):

> As of Kubernetes 1.8, you must specify a pod selector that matches the labels of the `.spec.template`. The pod selector will no longer be defaulted when left empty. Selector defaulting was not compatible with `kubectl apply`. Also, once a DaemonSet is created, its `.spec.selector` can not be mutated. Mutating the pod selector can lead to the unintentional orphaning of Pods, and it was found to be confusing to users.

ChiefAlexander replied:
Make the selectors only be `app` and `release` which is pretty standard. As @taharah mentioned it is not a good idea to include the custom podLabels here.

taharah commented on stable/nginx-ingress/templates/controller-deployment.yaml:
```diff
@@ -1,5 +1,5 @@
 {{- if or (eq .Values.controller.kind "Deployment") (eq .Values.controller.kind "Both") }}
-apiVersion: extensions/v1beta1
+apiVersion: apps/v1
```
Same as previous comment.

> I think this could potentially be a breaking change for some users, so can we just make this configurable with the default being the existing API version?

null93 replied:
The **Deployment** resource is no longer available in `v1.16` under the `extensions/v1beta1 ` API. It is now only available in the `apps/v1` API. This was available in the `apps/v1` API since `v1.9`. This should be backward compatible as long as `.spec.selector` is explicitly defined within the **Deployment** resource.

Source: https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16/

taharah commented on stable/nginx-ingress/templates/controller-deployment.yaml:
```diff
@@ -10,6 +10,14 @@ metadata:
+      component: "{{ .Values.controller.name }}"
+      release: {{ .Release.Name }}
+      {{- if .Values.controller.podLabels }}
+{{ toYaml .Values.controller.podLabels | indent 6 }}
```
Same as previous comment.

> Let's not add the podLabels to the selector since this field is immutable.

null93 replied:
According to [https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#selector](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#selector):

> `.spec.selector` is an required field that specifies a label selector for the Pods targeted by this Deployment.
>
> `.spec.selector` must match `.spec.template.metadata.labels`, or it will be rejected by the API.
> 
> In API version `apps/v1`, `.spec.selector` and `.metadata.labels` do not default to `.spec.template.metadata.labels` if not set. So they must be set explicitly. Also note that `.spec.selector` is immutable after creation of the Deployment in `apps/v1`.

taharah commented on stable/nginx-ingress/templates/controller-psp.yaml:
```diff
@@ -1,5 +1,5 @@
 {{- if .Values.podSecurityPolicy.enabled}}
-apiVersion: extensions/v1beta1
+apiVersion: policy/v1beta1
```
Same as previous comment.

> I think this could potentially be a breaking change for some users, so can we just make this configurable with the default being the existing API version?

null93 replied:
The **PodSecurityPolicy** resource is no longer available in `v1.16` under the `extensions/v1beta1` API. It is now only available in the `policy/v1beta1` API. This was available in the `policy/v1beta1` API since `v1.10`. This should be backward compatible, the API was just changed to a more stable endpoint.

Source: https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16/

taharah commented on stable/nginx-ingress/templates/default-backend-deployment.yaml:
```diff
@@ -1,5 +1,5 @@
 {{- if .Values.defaultBackend.enabled }}
-apiVersion: extensions/v1beta1
+apiVersion: apps/v1
```
Same as previous comment.

> I think this could potentially be a breaking change for some users, so can we just make this configurable with the default being the existing API version?

null93 replied:
The **Deployment** resource is no longer available in `v1.16` under the `extensions/v1beta1 ` API. It is now only available in the `apps/v1` API. This was available in the `apps/v1` API since `v1.9`. This should be backward compatible as long as `.spec.selector` is explicitly defined within the **Deployment** resource.

Source: https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16/

taharah replied:
Yeah we shouldn't have any issues since the selector field was not defined previously. I'd be comfortable having the default be `apps/v1`, but I would still like to see this be configurable so that users can continue to use an older API if they so desire.

null93 replied:
done :)

taharah commented on stable/nginx-ingress/templates/default-backend-deployment.yaml:
```diff
@@ -10,6 +10,14 @@ metadata:
+      component: "{{ .Values.defaultBackend.name }}"
+      release: {{ .Release.Name }}
+      {{- if .Values.defaultBackend.podLabels }}
+{{ toYaml .Values.defaultBackend.podLabels | indent 6 }}
```
Same as previous comment.

> Let's not add the podLabels to the selector since this field is immutable.

null93 replied:
According to [https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#selector](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#selector)

> `.spec.selector` is an required field that specifies a label selector for the Pods targeted by this Deployment.
>
> `.spec.selector` must match `.spec.template.metadata.labels`, or it will be rejected by the API.
> 
> In API version `apps/v1`, `.spec.selector` and `.metadata.labels` do not default to `.spec.template.metadata.labels` if not set. So they must be set explicitly. Also note that `.spec.selector` is immutable after creation of the Deployment in `apps/v1`.

taharah replied:
Since the field is immutable we want to only use the minimum necessary amount of labels to uniquely identify the pods. Otherwise, there is a large potential to not be able to do upgrades. For instance, if I add a `version` label, then I will always need to re-create the deployment in order to upgrade.

taharah reviewed (changes requested):
Overall, this looks fine and I'm happy with the changes. I just want to try and ensure that we do not introduce any potential backwards incompatibilities.

null93 reviewed (commented):
Please see the comments. Should be backward compatible to my knowledge. Sources were linked alongside comments.

Commit:
[stable/nginx-ingress] Fixes compatibility with 1.16

Notice of deprecation: https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16

Signed-off-by: Rafael Grigorian <me@raffi.io>

Commit:
Only kept 'app' and 'release' in selectors

Signed-off-by: Rafael Grigorian <me@raffi.io>
